File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int[] max_till_now = nums.clone();
        int[] min_till_now = nums.clone();
        int[] res = nums.clone();
        
        for(int i=1; i<nums.length; i++) {
            min_till_now[i] = Math.min(nums[i], Math.min(nums[i]*min_till_now[i-1], nums[i]*max_till_now[i-1]));
            max_till_now[i] = Math.max(nums[i], Math.max(nums[i]*min_till_now[i-1], nums[i]*max_till_now[i-1]));
            
            res[i] = Math.max(res[i-1], max_till_now[i]);
            
        }
        return res[nums.length - 1];
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    // 2 loop approach 
    /*int maxProduct(vector<int>& nums) {
        
        int maxLeft = INT_MIN;
        int maxRight = INT_MIN;
        
        // curr product
        int currProd = 1;
        
        // flag to check the presence of 0
        bool zeroPresent =  false;
        
        for(int i=0; i<nums.size()-1; i++) {
            // if curr elem is 0, reset the product, mark the zero flag and go to next iteration
            if(nums[i] == 0) {
                zeroPresent = true;
                currProd = 1;
                continue;
            }
            // else just update curr prod and lmax
            currProd *= nums[i];
            maxLeft = max(maxLeft,currProd);
        }
        
        currProd = 1;
        
        for(int j=nums.size()-1; j>=0; j--) {
            // if curr elem is 0, reset the product, mark the zero flag and go to next iteration
            if(nums[j] == 0) {
                zeroPresent = true;
                currProd = 1;
                continue;
            }
            // else just update curr prod and rmax
            currProd *= nums[j];
            maxRight = max(maxRight,currProd);
        }
        
        if(zeroPresent) {
            return max(max(maxLeft,maxRight),0);
        }

        return max(maxLeft,maxRight);
        
    }*/

    // 1 loop approach
    int maxProduct(vector<int>& nums) {
        int minProd = nums[0];
        int maxProd = nums[0];
        int result = nums[0];

        for(int i=1; i<nums.size(); i++) {
            // first store maxProd's final value in a temp variable so that it doesn't get affected when we update minProd
            int temp = max({nums[i], nums[i] * maxProd, nums[i] * minProd});
            minProd = min({nums[i], nums[i] * maxProd, nums[i] * minProd});
            maxProd = temp;

            result = max(result, maxProd);

        }
        return result;
    }

};